rust/lance/src/dataset/branch_location.rs

```diff
@@ -0,0 +1,191 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: Copyright The Lance Authors
+
+use lance_core::{Error, Result};
+use object_store::path::Path;
+use snafu::location;
+
+#[derive(Debug, Clone, PartialEq, Hash)]
+pub struct BranchLocation {
+    pub path: Path,
+    pub uri: String,
+    pub branch: Option<String>,
+}
+
+impl BranchLocation {
+    /// Find the root location
+    pub fn find_root(&self) -> Result<Self> {
+        if let Some(branch_name) = self.branch.as_ref() {
+            let mut path = self.path.clone();
+            let mut uri = self.uri.clone();
+            let segment_count = branch_name.split('/').count();
+            for _ in 0..segment_count + 1 {
+                path = Path::parse(Self::parent_str(path.as_ref())?)?;
+                uri = Self::parent_str(uri.as_str())?.to_string();
+            }
+            Ok(Self {
+                path,
+                uri,
+                branch: None,
+            })
+        } else {
+            Ok(self.clone())
+        }
+    }
+
+    /// Find the target branch location
+    pub fn find_branch(&self, branch_name: Option<String>) -> Result<Self> {
+        if branch_name == self.branch {
+            return Ok(self.clone());
+        }
+
+        let root_location = self.find_root()?;
+        if let Some(target_branch) = branch_name.as_ref() {
+            let (new_path, new_uri) = {
+                // Handle empty segment
+                if target_branch.is_empty() {
+                    (self.path.clone(), self.uri.clone())
+                } else {
+                    let segments = target_branch.split('/');
+                    let mut new_path_str = Self::join_str(root_location.path.as_ref(), "tree")?;
+                    let mut new_uri = Self::join_str(root_location.uri.as_str(), "tree")?;
+                    for segment in segments {
+                        new_path_str = Self::join_str(new_path_str.as_str(), segment)?;
+                        new_uri = Self::join_str(new_uri.as_str(), segment)?;
+                    }
+                    (Path::parse(new_path_str)?, new_uri)
+                }
+            };
+            Ok(Self {
+                path: new_path,
+                uri: new_uri,
+                branch: Some(target_branch.clone()),
+            })
+        } else {
+            Ok(root_location)
+        }
+    }
+
+    fn parent_str(path_str: &str) -> Result<&str> {
+        let trimmed = path_str.trim_end_matches('/');
+        match trimmed.rfind('/') {
```

Comment:
would this fail on windows?

Reply:
why is that?

Reply:
Do we in general support paths like `C:\Users\Username\Documents\file`? Maybe I am just over thinking.